hls-plugin-api/src/Ide/Types.hs

```diff
@@ -693,6 +693,24 @@ nullToMaybe' :: (a |? (b |? Null)) -> Maybe (a |? b)
 nullToMaybe' (InL x)       = Just $ InL x
 nullToMaybe' (InR (InL x)) = Just $ InR x
 nullToMaybe' (InR (InR _)) = Nothing
+
+type Definitions = (Definition |? ([DefinitionLink] |? Null))
+
+mergeDefinitions :: Definitions -> Definitions -> Definitions
+mergeDefinitions definitions1 definitions2 = case (definitions1, definitions2) of
+    (InR (InR Null), def2) -> def2
+    (def1, InR (InR Null)) -> def1
+    (InL def1, InL def2) -> InR $ InL (defToLinks def1 ++ defToLinks def2)
+    (InL def1, InR (InL links)) -> InR $ InL (defToLinks def1 ++ links)
+    (InR (InL links), InL def2) -> InR $ InL (links ++ defToLinks def2)
+    (InR (InL links1), InR (InL links2)) -> InR $ InL (links1 ++ links2)
+    where
+        defToLinks :: Definition -> [DefinitionLink]
+        defToLinks (Definition (InL location)) = [DefinitionLink $ locationToLocationLink location]
+        defToLinks (Definition (InR locations)) = map (DefinitionLink . locationToLocationLink) locations
+
+        locationToLocationLink :: Location -> LocationLink
```

Comment:
Nitpick

How about:

```suggestion
        locationToDefinitionLink :: Location -> DefinitionLink
```

Since you use `locationToLocationLink` exclusively to create a `DefinitionLink`.